### PR TITLE
Enhance AAD Provider and Rego Code to automate checks for policy MS.AAD.3.3v1

### LINKED ISSUE #410: Implement new AAD Rego policy MS.AAD.3.3v1 MS authenticator context information
# 💡 Summary #

Implement a new Rego policy check for MS.AAD.3.3v1 If Phishing-resistant MFA has not been enforced yet and Microsoft Authenticator is enabled, it SHALL be configured to show login context information.

- If policy MS.AAD.3.1v1 Passes, then this policy MS.AAD.3.3v1 will automatically Pass. See the way the OneDrive expiration date for Anyone links is dependent on a different setting for an example.
- You must also check if MS Authenticator is enabled on the Azure Active Directory > Security > Authentication methods page. If Authenticator is not enabled, then this policy will automatically Pass since there is no reason to check the configuration settings.
- If Authenticator is enabled, then you need to check to ensure the [Authenticator settings match the values detailed in the implementation instructions](https://github.com/cisagov/ScubaGear/blob/emerald/baselines/aad.md#msaad33v1-instructions).

## Acceptance criteria ##

How do we know when this work is done?

- [ ] Test the code for all possible value combinations of MS.AAD.3.1v1 and MS.AAD.3.3v1 to ensure that the report produces an accurate Pass or Fail. You must also test scenarios where the MS Authenticator authentication method is Enabled and Disabled to ensure the results product expected evaluations.
- [ ] Test in all tenants

### LINKED ISSUE #1020: AAD policy 3.5 should produce a not applicable message when the authentication policy migration state is not set to Migration Complete
# 🐛 Summary #

AAD policy 3.5 is written incorrectly for the scenario when the tenant's authentication method policy migration state is not set to migrationComplete. As per the policy note in the baseline, 3.5 is only applicable "if the tenant has their Manage Migration feature set to Migration Complete" but it is currently producing a vanilla Fail in that scenario. Instead it should produce a Shall/Not-Implemented with the message "This policy is only applicable if the tenant has their Manage Migration feature set to Migration Complete."

## To reproduce ##

1. Run ScubaGear to export the provider settings.
2. Copy the ProviderSettingsExport.json file to a Testing folder
3. Open the file and set the PolicyMigrationState value to preMigration.
4. Run ScubaGear using the run cached cmdlet below
`Invoke-RunCached -ProductNames aad -ExportProvider $false -OutPath C:\Users\tkolovos\Downloads\Testing`
6. Examine the report and you will see that it incorrectly produces a Fail

image

## Implementation

- [ ] Fix the Rego policy so that it produces a Shall/Not-Implemented with the message "This policy is only applicable if the tenant has their Manage Migration feature set to Migration Complete."
- [ ] Update the functional test "MS.AAD.3.5v1 Non-compliant case - preMigration" so that it is named "Not applicable case - preMigration" and set the state of SMS to enabled.

### PR DESCRIPTION
**Pull request by ssatyapal123**
#  #
  

## 🗣 Description ##
Updated AAD Rego code  to check for MS.AAD.3.3v1: If Phishing-resistant MFA has not been enforced yet and Microsoft Authenticator is enabled, it SHALL be configured to show login context information.

Closes #410 
Closes #1020 

Changes in detail:

- To code this issue for policy 3.3, we needed to retrieve the authentication method configurations that were already being returned by an existing ScubaGear cmdlet Get-MgBetaPolicyAuthenticationMethodPolicy for policies 3.4 & 3.5 but the JSON structure had to be modified to support all three policies.
- The Rego was updated for policies 3.4 and 3.5 to read from the adjusted JSON structure. Unit tests and functional tests were adjusted accordingly.
- Rego policy 3.5 was adjusted since it was coded wrong - when the policy migration state is not migrationComplete it should have a Criticality of Shall/Not-Implemented to align with the baseline requirement. This results in an N/A in the report Result column.
- Products.Tests.ps1 was modified to support the case when a policy (e.g. 3.3) is N/A but has a tailored Details message instead of the generic "This product does not currently have the capability to check compliance for this policy" which only makes sense in some scenarios but not for policy 3.3.
- The Rego for new policy implementation 3.3 was coded. Unit tests and functional tests were coded as well.

## 💭 Motivation and context ##
Previously MSAAD.3.1v1 was not being checked by ScubaGear. This change adds automation by checking if MS Authenticator has the appropriate context information enabled.

## 🧪 Testing ##
Tested against E5/G5 tenant to confirm 3.3v1 passes/fails when the settings for Microsoft Authenticator are changed.

## ✅ Pre-approval checklist ##

- [x] This PR has an informative and human-readable title.
- [x] PR targets the correct parent branch (e.g., main or release-name) for merge.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] Changes are sized such that they do not touch excessive number of files.
- [x] *All* future TODOs are captured in issues, which are referenced in code comments.
- [x] These code changes follow the ScubaGear [content style guide](https://github.com/cisagov/ScubaGear/blob/main/CONTENTSTYLEGUIDE.md).
- [x] Related issues these changes resolve are linked preferably via [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).
- [x] All relevant type-of-change labels added.
- [x] All relevant project fields are set.
- [x] All relevant repo and/or project documentation updated to reflect these changes.
- [x] Unit tests added/updated to cover PowerShell and Rego changes.
- [x] Functional tests added/updated to cover PowerShell and Rego changes.
- [x] All relevant functional tests passed.
- [ ] All automated checks (e.g., linting, static analysis, unit/smoke tests) passed.

## ✅ Pre-merge checklist ##

- [ ] PR passed smoke test check.
- [ ] Feature branch has been rebased against changes from parent branch, as needed

  Use `Rebase branch` button below or use [this](https://www.digitalocean.com/community/tutorials/how-to-rebase-and-update-a-pull-request) reference to rebase from the command line.
- [ ] Resolved all merge conflicts on branch
- [ ] Notified merge coordinator that PR is ready for merge via comment mention

## ✅ Post-merge checklist ##

- [ ] Feature branch deleted after merge to clean up repository.
- [ ] Verified that all checks pass on parent branch (e.g., main or release-name) after merge.

### CODE REVIEW

**Comment by tkol2022** [comment]
@ssatyapal123 

- You need to go through the pre-approval checklist at the top of the PR
- Your branch is out of date with the main branch. Rebase.
- Functional tests are missing so create a separate issue and we will add those in the next release

image

File: PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1
Comment (james-garriss) [inline PowerShell/ScubaGear/Modules/Providers/ExportAADProvider.psm1:1]: Appreciate the many comments.

File: PowerShell/ScubaGear/Rego/AADConfig.rego
Comment (james-garriss) [inline PowerShell/ScubaGear/Rego/AADConfig.rego:1]: Updated variable names track.

File: Testing/Functional/Products/Products.Tests.ps1
Comment (james-garriss) [inline Testing/Functional/Products/Products.Tests.ps1:1]: When we say, "A custom product can be used to fulfill this policy requirement", do we provide info anywhere on how the user can do this?  Maybe it's in the readme somewhere?

**Review by james-garriss — APPROVED** [review]
Nice work.

File: Testing/Functional/Products/Products.Tests.ps1
Comment (tkol2022) [reply Testing/Functional/Products/Products.Tests.ps1:1]: I believe this is meant to describe the case where a third party product can be used to implement the policy and there is no Scuba test for that specific policy. See below for an example from EXO. AAD functional tests do not use the IsCustomImplementation field.

image

image

**Comment by tkol2022** [comment]
> Nice work.

thx

**Review by dagarwal-mitre — APPROVED** [review]

**Comment by tkol2022** [comment]
@nanda-katikaneni - ready for merge

**Pull request merged**